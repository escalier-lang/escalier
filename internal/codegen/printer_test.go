package codegen

import (
	"testing"

	"github.com/escalier-lang/escalier/internal/parser"
	"github.com/gkampitakis/go-snaps/snaps"
)

func TestPrintExpr(t *testing.T) {
	sum := &BinaryExpr{
		Left: &NumExpr{
			Value:  0.1,
			span:   nil,
			source: nil,
		},
		Op: Plus,
		Right: &NumExpr{
			Value:  0.2,
			span:   nil,
			source: nil,
		},
		span:   nil,
		source: nil,
	}

	printer := NewPrinter()
	printer.PrintExpr(sum)

	want := "0.1 + 0.2"
	if got := printer.Output; got != want {
		t.Errorf("got %q, want %q", got, want)
	}
	snaps.MatchSnapshot(t, sum)
}

func TestPrintModule(t *testing.T) {
	source := parser.Source{
		Path: "input.esc",
		Contents: `fn add(a, b) { return a + b }
fn sub(a, b) { return a - b }`,
	}
	p := parser.NewParser(source)
	m1 := p.ParseModule()
	builder := &Builder{
		tempId: 0,
	}
	m2 := builder.BuildModule(m1)

	printer := NewPrinter()
	printer.PrintModule(m2)

	snaps.MatchSnapshot(t, printer.Output)
	if printer.location.Line != 11 {
		t.Errorf("got %d, want 11", printer.location.Line)
	}
}
