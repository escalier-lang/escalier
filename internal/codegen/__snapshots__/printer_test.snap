
[TestPrintExpr - 1]
&codegen.BinaryExpr{
    Left: &codegen.LitExpr{
        Lit: &codegen.NumLit{
            Value:  0.1,
            span:   (*codegen.Span)(nil),
            source: nil,
        },
        span: &codegen.Span{
            Start: codegen.Location{Line:1, Column:1},
            End:   codegen.Location{Line:1, Column:4},
        },
        source: nil,
    },
    Op:    "+",
    Right: &codegen.LitExpr{
        Lit: &codegen.NumLit{
            Value:  0.2,
            span:   (*codegen.Span)(nil),
            source: nil,
        },
        span: &codegen.Span{
            Start: codegen.Location{Line:1, Column:7},
            End:   codegen.Location{Line:1, Column:10},
        },
        source: nil,
    },
    span: &codegen.Span{
        Start: codegen.Location{Line:1, Column:1},
        End:   codegen.Location{Line:1, Column:10},
    },
    source: nil,
}
---

[TestPrintModule - 1]
function add(temp1, temp2) {
  const a = temp1;
  const b = temp2;
  return a + b;
}
function sub(temp3, temp4) {
  const a = temp3;
  const b = temp4;
  return a - b;
}

---
